# Tax-Min-Method

In [54]:
pip install pandas-datareader

You should consider upgrading via the '/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [55]:
#Import required libraries
import pandas as pd
import numpy as np
from datetime import date
import decimal
from pandas_datareader import data as pdr
from datetime import datetime

In [56]:
import os
working_directory = os.getcwd()

## Import CSV file and convert "Date Acquired" column to datetime 

In [57]:
path = working_directory + '/data/Min_Tax_Port.csv'
d_parser = lambda x: datetime.strptime(x, '%m/%d/%Y')
tax_min_port = pd.read_csv(path, parse_dates=["Date Acquired"], date_parser=d_parser)
tax_min_port

Symbol  Quanity  Last Price  Cost/Share  Total Gain Date Acquired
0     VXUS  40.0000       65.59     63.6300     78.4000    2021-02-24
1     VXUS  25.0000       65.59     63.1275     61.5625    2021-03-12
2     VXUS   0.1637       65.59     61.6282      0.6486    2021-03-25
3     VXUS   0.5079       65.59     66.0598     -0.2386    2021-06-24
4      VWO  55.0000       51.64     54.9061   -179.6355    2021-02-22
..     ...      ...         ...         ...         ...           ...
102  ABALX  36.5680       33.13     29.8800    118.8478    2020-12-16
103  ABALX   4.7680       33.13     29.8784     15.5038    2020-12-16
104  ABALX   4.7100       33.13     31.1231      9.4523    2021-03-16
105  ABALX   8.0030       33.13     32.7090      3.3694    2021-06-15
106  ABALX   4.4960       33.13     32.7091      1.8925    2021-06-15

[107 rows x 6 columns]

# Create variables that gives us the current date and date one year prior.
This logic is to determine the current date and then the date one year prior which would indicate any lots purchased before then (the date one year prior from current date) are definitely long term tax lots.
Some brokerage firms platforms will actually allow you to define which tax lots are long or short, however, that is too easy and defining a function or method to determine this will be helpful for any data that does not identify the tax lots beforehand.

In [59]:
import datetime
#date_now = datetime.datetime.now()
date_now = datetime.date.today()
year_ago = date_now.year - 1

current_date = date_now.strftime('%Y-%m-%d')
one_year_ago = date_now.replace(year=year_ago).strftime('%Y-%m-%d')

print(current_date)
print(one_year_ago)

2021-08-15
2020-08-15


## Filter the dataframe to create two new datasets that specify all Long-Term lots and all Short-Term Lots
These are not really relevant for this project but just wanted to filter for good measure.

In [60]:
# Filter dataframe to show all tax lots that are considered long term gains/losses
long_term_lots = (tax_min_port['Date Acquired'] <= one_year_ago)
long_term = tax_min_port.loc[long_term_lots]
long_term

Symbol    Quanity  Last Price  Cost/Share  Total Gain Date Acquired
10     VTV    23.0000      141.19    120.6374    472.7100    2020-02-16
11     VTV    27.0000      141.19     96.9500   1194.4800    2020-05-22
16     VTI    41.8022      228.93    121.3919   4495.3277    2020-03-24
17     VTI     0.1978      228.93    130.2831     19.5124    2020-03-30
29    TDOC     5.1200      146.79     65.1520    417.9864    2020-05-27
30    TDOC     8.8800      146.79    158.5980   -104.8548    2020-07-15
32       T   100.0000       28.09     29.9169   -182.6900    2020-07-20
39     SPG    10.0000      132.11     62.5000    696.1000    2020-05-27
40     SPG    15.0000      132.11     60.7593   1070.2600    2020-07-20
45    ROKU    15.0000      379.26    125.0000   3813.9000    2020-06-19
51     IJS    25.9399      102.24     77.6605    637.5905    2020-02-16
62    FSLY    10.0000       42.30     78.6700   -363.7000    2020-07-16
63    FAST   109.4789       54.87     38.2212   1822.6873    2020-02-16
64    FAST     1.5211       54.87     36.5262     27.9028    2020-02-26
69    DKNG    25.0000       51.98     39.8500    303.2500    2020-06-23
70    DKNG    25.0000       51.98     35.0000    424.5000    2020-06-25
72     CWH    30.0000       40.73     34.5000    186.9000    2020-08-06
87   AMECX  2230.3520       26.10     23.5026   5793.0672    2020-02-16
88   AMECX    20.1540       26.10     23.5025     52.3494    2020-03-16
89   AMECX    17.6490       26.10     21.0397     89.3089    2020-06-15
96   ABALX  1401.2280       33.13     29.3268   5329.1139    2020-02-16
97   ABALX     5.8850       33.13     29.3254     22.3900    2020-03-16
98   ABALX     0.4840       33.13     27.6240      2.6649    2020-06-15
99   ABALX     5.0910       33.13     27.6390     27.9548    2020-06-15
100  ABALX     7.0000       33.13     27.6400     38.4300    2020-06-15

In [61]:
# Filter dateframe to include tax lots that are considered short term gains/losses
short_term_lots = (tax_min_port['Date Acquired'] > one_year_ago)
short_term = tax_min_port.loc[short_term_lots]
short_term

Symbol  Quanity  Last Price  Cost/Share  Total Gain Date Acquired
0     VXUS  40.0000       65.59     63.6300     78.4000    2021-02-24
1     VXUS  25.0000       65.59     63.1275     61.5625    2021-03-12
2     VXUS   0.1637       65.59     61.6282      0.6486    2021-03-25
3     VXUS   0.5079       65.59     66.0598     -0.2386    2021-06-24
4      VWO  55.0000       51.64     54.9061   -179.6355    2021-02-22
..     ...      ...         ...         ...         ...           ...
102  ABALX  36.5680       33.13     29.8800    118.8478    2020-12-16
103  ABALX   4.7680       33.13     29.8784     15.5038    2020-12-16
104  ABALX   4.7100       33.13     31.1231      9.4523    2021-03-16
105  ABALX   8.0030       33.13     32.7090      3.3694    2021-06-15
106  ABALX   4.4960       33.13     32.7091      1.8925    2021-06-15

[82 rows x 6 columns]

## Now filter the original dataframe to display four separate categories.
 1.) Short-Term losses 2.) Long-Term losses 3.) Long-Term gains 4.) Short-Term gains.

These dataframes will also be filtered to display the lots with the highest cost bases (Cost/Share) in descending order.

In [62]:
#Filter dataframe to display short term losses and highest cost basis in descending order
short = (tax_min_port['Total Gain'] < 0) & (tax_min_port['Date Acquired'] > one_year_ago)
short_term_loss = tax_min_port.loc[short].sort_values(by='Cost/Share', ascending=False)
short_term_loss

Symbol   Quanity  Last Price  Cost/Share  Total Gain Date Acquired
28    VBK    0.0104      284.37    291.9000     -0.0784    2021-06-29
31   TDOC   11.0000      146.79    199.0000   -574.3100    2021-01-05
54    IJS    0.0606      102.24    107.0604     -0.2922    2021-06-16
86    BND    0.0553       85.95     86.7399     -0.0437    2021-08-05
85    BND    0.0539       85.95     86.3850     -0.0235    2021-07-07
80    BND   35.0000       85.95     86.0000     -1.7500    2021-02-22
3    VXUS    0.5079       65.59     66.0598     -0.2386    2021-06-24
4     VWO   55.0000       51.64     54.9061   -179.6355    2021-02-22
6     VWO    0.2858       51.64     54.0164     -0.6793    2021-06-24
50    NIO   25.0000       44.22     48.4078   -104.6950    2021-02-23
35      T    1.7035       28.09     31.6697     -6.0981    2021-05-03
34      T    1.8434       28.09     28.7463     -1.2098    2021-02-01
36      T    1.9324       28.09     28.3793     -0.5590    2021-08-02
94  AMECX   14.6960       26.10     26.1105     -0.1544    2021-06-15
48   PLTR  100.0000       22.92     25.9800   -306.0000    2021-01-12
49   PLTR  100.0000       22.92     25.0000   -208.0000    2021-02-25

In [63]:
#Filter dataframe to display long term losses and highest cost basis in descending order
long = (tax_min_port['Total Gain'] < 0) & (tax_min_port['Date Acquired'] <= one_year_ago)
long_term_loss = tax_min_port.loc[long].sort_values(by='Cost/Share', ascending=False)
long_term_loss

Symbol  Quanity  Last Price  Cost/Share  Total Gain Date Acquired
30   TDOC     8.88      146.79    158.5980   -104.8548    2020-07-15
62   FSLY    10.00       42.30     78.6700   -363.7000    2020-07-16
32      T   100.00       28.09     29.9169   -182.6900    2020-07-20

In [64]:
#Filter dataframe to display long term gains and highest cost basis in descending order
long = (tax_min_port['Total Gain'] > 0) & (tax_min_port['Date Acquired'] <= one_year_ago)
long_term_gain = tax_min_port.loc[long].sort_values(by='Cost/Share', ascending=False)
long_term_gain

Symbol    Quanity  Last Price  Cost/Share  Total Gain Date Acquired
17     VTI     0.1978      228.93    130.2831     19.5124    2020-03-30
45    ROKU    15.0000      379.26    125.0000   3813.9000    2020-06-19
16     VTI    41.8022      228.93    121.3919   4495.3277    2020-03-24
10     VTV    23.0000      141.19    120.6374    472.7100    2020-02-16
11     VTV    27.0000      141.19     96.9500   1194.4800    2020-05-22
51     IJS    25.9399      102.24     77.6605    637.5905    2020-02-16
29    TDOC     5.1200      146.79     65.1520    417.9864    2020-05-27
39     SPG    10.0000      132.11     62.5000    696.1000    2020-05-27
40     SPG    15.0000      132.11     60.7593   1070.2600    2020-07-20
69    DKNG    25.0000       51.98     39.8500    303.2500    2020-06-23
63    FAST   109.4789       54.87     38.2212   1822.6873    2020-02-16
64    FAST     1.5211       54.87     36.5262     27.9028    2020-02-26
70    DKNG    25.0000       51.98     35.0000    424.5000    2020-06-25
72     CWH    30.0000       40.73     34.5000    186.9000    2020-08-06
96   ABALX  1401.2280       33.13     29.3268   5329.1139    2020-02-16
97   ABALX     5.8850       33.13     29.3254     22.3900    2020-03-16
100  ABALX     7.0000       33.13     27.6400     38.4300    2020-06-15
99   ABALX     5.0910       33.13     27.6390     27.9548    2020-06-15
98   ABALX     0.4840       33.13     27.6240      2.6649    2020-06-15
87   AMECX  2230.3520       26.10     23.5026   5793.0672    2020-02-16
88   AMECX    20.1540       26.10     23.5025     52.3494    2020-03-16
89   AMECX    17.6490       26.10     21.0397     89.3089    2020-06-15

In [65]:
#Filter dataframe to display short term gain
short = (tax_min_port['Total Gain'] > 0) & (tax_min_port['Date Acquired'] > one_year_ago)
short_term_gain = tax_min_port.loc[short].sort_values(by='Cost/Share', ascending=False)
short_term_gain

Symbol   Quanity  Last Price  Cost/Share  Total Gain Date Acquired
9     VUG    0.0131      297.10    286.3999      0.1397    2021-06-29
25    VBK    5.0000      284.37    281.4320     14.6900    2021-03-17
26    VBK    5.0000      284.37    274.7083     48.3085    2021-03-18
57    IJH    0.0137      271.49    270.0590      0.0196    2021-06-16
27    VBK    0.0072      284.37    266.5698      0.1276    2021-03-30
..    ...       ...         ...         ...         ...           ...
93  AMECX   15.4140       26.10     24.7295     21.1254    2021-03-16
92  AMECX    9.0030       26.10     23.3500     24.7583    2020-12-16
91  AMECX   16.1470       26.10     23.3500     44.4042    2020-12-16
90  AMECX   16.8810       26.10     22.1700     66.3423    2020-09-15
79    CGC  100.0000       19.47     18.8500     62.0000    2020-10-20

[66 rows x 6 columns]

## Identify the total gain or loss for each new category. No practical use for this excercise, just helpful information to see.

In [66]:
short_term_loss['Total Gain'].sum()

-1383.7675

In [67]:
long_term_loss['Total Gain'].sum()

-651.2447999999999

In [68]:
long_term_gain['Total Gain'].sum()

26938.386199999997

In [69]:
short_term_gain['Total Gain'].sum()

8704.2762

## Exhaust each category before moving to the next, but within each category, lots with the highest cost basis are sold first.
Will also append another column to each category to include the total proceeds from each lot. (Last Price * Quanity = Total Value).  Although for this section only the Short-Term Loss category is being displayed.

In [70]:
short_term_loss["Total Value"] = (short_term_loss["Last Price"] * short_term_loss["Quanity"])

In [71]:
short_term_loss

Symbol   Quanity  Last Price  Cost/Share  Total Gain Date Acquired  \
28    VBK    0.0104      284.37    291.9000     -0.0784    2021-06-29   
31   TDOC   11.0000      146.79    199.0000   -574.3100    2021-01-05   
54    IJS    0.0606      102.24    107.0604     -0.2922    2021-06-16   
86    BND    0.0553       85.95     86.7399     -0.0437    2021-08-05   
85    BND    0.0539       85.95     86.3850     -0.0235    2021-07-07   
80    BND   35.0000       85.95     86.0000     -1.7500    2021-02-22   
3    VXUS    0.5079       65.59     66.0598     -0.2386    2021-06-24   
4     VWO   55.0000       51.64     54.9061   -179.6355    2021-02-22   
6     VWO    0.2858       51.64     54.0164     -0.6793    2021-06-24   
50    NIO   25.0000       44.22     48.4078   -104.6950    2021-02-23   
35      T    1.7035       28.09     31.6697     -6.0981    2021-05-03   
34      T    1.8434       28.09     28.7463     -1.2098    2021-02-01   
36      T    1.9324       28.09     28.3793     -0.5590    2021-08-02   
94  AMECX   14.6960       26.10     26.1105     -0.1544    2021-06-15   
48   PLTR  100.0000       22.92     25.9800   -306.0000    2021-01-12   
49   PLTR  100.0000       22.92     25.0000   -208.0000    2021-02-25   

    Total Value  
28     2.957448  
31  1614.690000  
54     6.195744  
86     4.753035  
85     4.632705  
80  3008.250000  
3     33.313161  
4   2840.200000  
6     14.758712  
50  1105.500000  
35    47.851315  
34    51.781106  
36    54.281116  
94   383.565600  
48  2292.000000  
49  2292.000000

This is where it gets tricky.  Consider a one-time $10,000 withdrawal.  Looking at the previous dataframe I can tell that we will only need to iterate through the first category (Short Term Loss) to get to this value.

Below will display in essence a new proposal of all the tax lots that should be sold to realize short term losses from the highest cost basis first.  However,  I need to figure out a way for the proposal to consider if a partial lot need be sold to achieve the total withdrawal amount.  Additonally, I need to write a function or loop that will be able to iterate through the remaining categories in order if applicable.

In [72]:
# Consider a total withdrawal amount and write a function that 
withdrawal_amount = 10000

In [73]:
recommended_tax_lots = short_term_loss.loc[short_term_loss['Total Value'].cumsum().le(withdrawal_amount)]

In [74]:
recommended_tax_lots

Symbol  Quanity  Last Price  Cost/Share  Total Gain Date Acquired  \
28    VBK   0.0104      284.37    291.9000     -0.0784    2021-06-29   
31   TDOC  11.0000      146.79    199.0000   -574.3100    2021-01-05   
54    IJS   0.0606      102.24    107.0604     -0.2922    2021-06-16   
86    BND   0.0553       85.95     86.7399     -0.0437    2021-08-05   
85    BND   0.0539       85.95     86.3850     -0.0235    2021-07-07   
80    BND  35.0000       85.95     86.0000     -1.7500    2021-02-22   
3    VXUS   0.5079       65.59     66.0598     -0.2386    2021-06-24   
4     VWO  55.0000       51.64     54.9061   -179.6355    2021-02-22   
6     VWO   0.2858       51.64     54.0164     -0.6793    2021-06-24   
50    NIO  25.0000       44.22     48.4078   -104.6950    2021-02-23   
35      T   1.7035       28.09     31.6697     -6.0981    2021-05-03   
34      T   1.8434       28.09     28.7463     -1.2098    2021-02-01   
36      T   1.9324       28.09     28.3793     -0.5590    2021-08-02   
94  AMECX  14.6960       26.10     26.1105     -0.1544    2021-06-15   

    Total Value  
28     2.957448  
31  1614.690000  
54     6.195744  
86     4.753035  
85     4.632705  
80  3008.250000  
3     33.313161  
4   2840.200000  
6     14.758712  
50  1105.500000  
35    47.851315  
34    51.781106  
36    54.281116  
94   383.565600

In [75]:
recommended_tax_lots['Total Value'].sum()

9172.729942

In [76]:
proceeds_still_needed = withdrawal_amount - recommended_tax_lots['Total Value'].sum()
proceeds_still_needed

827.2700580000001

In this instance above, the proposal still requires $827.27 to meet the $10,000 withdrawal amount.  When viewing the next lot in the category, 36.09 shares of PLTR from the 2021-01-12 tax lot would need to be sold to achieve this.

How do we account for partial lots in the proposal???

The functions below are just for informational purposes.  acumsum reflects the total value in descending order from each lot.  The second just determines the first index where the value is greater than 10k.

In [77]:
# test = short_term_loss.loc[short_term_loss.sort_values('Total Proceeds',ascending=False,ignore_index=True)['Total Proceeds'].cumsum().le(10000)]

In [78]:
acumsum = np.cumsum(short_term_loss["Total Value"])

In [51]:
acumsum

28        2.957448
31     1617.647448
54     1623.843192
86     1628.596227
85     1633.228932
80     4641.478932
3      4674.792093
4      7514.992093
6      7529.750805
50     8635.250805
35     8683.102120
34     8734.883226
36     8789.164342
94     9172.729942
48    11464.729942
49    13756.729942
Name: Total Value, dtype: float64

In [52]:
np.argmax(acumsum > withdrawal_amount)

14